buildserver/deps.go

```diff
@@ -0,0 +1,500 @@
+		defer func() { <-gate }()
+
+		mu := importPathMutex.get(path)
+		mu.Lock() // only try to import a path once
+		defer mu.Unlock()
+
+		dc.importCacheMu.Lock()
+		res, ok = dc.importCache[importKey{path, srcDir, mode}]
+		dc.importCacheMu.Unlock()
+		if !ok {
+			res.pkg, res.err = importPackage(path, srcDir, mode)
+			dc.importCacheMu.Lock()
+			dc.importCache[importKey{path, srcDir, mode}] = res
+			dc.importCacheMu.Unlock()
+		}
+		return res.pkg, res.err
+	}
+
+	var errs errorList
+	var wg sync.WaitGroup
+	var do func(pkg *build.Package)
+	do = func(pkg *build.Package) {
+		dc.seenMu.Lock()
+		if _, seen := dc.seen[pkg.Dir]; seen {
+			dc.seenMu.Unlock()
+			return
+		}
+		dc.seen[pkg.Dir] = []importRecord{}
+		dc.seenMu.Unlock()
+
+		for _, path := range allPackageImportsSorted(pkg) {
+			if path == "C" {
+				continue
+			}
+			wg.Add(1)
+			parentPkg := pkg
+			go func(path string) {
+				defer wg.Done()
+				pkg, err := cachedImportPackage(path, pkg.Dir, mode)
+				if err != nil {
+					errs.add(err)
+				}
+				if pkg != nil {
+					if dc.collectReferences {
+						dc.seenMu.Lock()
+						dc.seen[parentPkg.Dir] = append(dc.seen[parentPkg.Dir], importRecord{pkg: parentPkg, imports: pkg})
+						dc.seenMu.Unlock()
+					}
+					do(pkg)
+				}
+			}(path)
+		}
+	}
+	do(pkg)
+	if dc.collectReferences {
+		dc.seenMu.Lock()
+		dc.entryPackageDirs = append(dc.entryPackageDirs, pkg.Dir)
+		dc.seenMu.Unlock()
+	}
+	wg.Wait()
+	return errs.error()
+}
+
+func allPackageImportsSorted(pkg *build.Package) []string {
+	uniq := map[string]struct{}{}
+	for _, p := range pkg.Imports {
+		uniq[p] = struct{}{}
+	}
+	for _, p := range pkg.TestImports {
+		uniq[p] = struct{}{}
+	}
+	for _, p := range pkg.XTestImports {
+		uniq[p] = struct{}{}
+	}
+	imps := make([]string, 0, len(uniq))
+	for p := range uniq {
+		imps = append(imps, p)
+	}
+	sort.Strings(imps)
+	return imps
+}
+
+func isMultiplePackageError(err error) bool {
+	_, ok := err.(*build.MultiplePackageError)
+	return ok
+}
+
+// FetchCommonDeps will fetch our common used dependencies. This is to avoid
+// impacting the first ever typecheck we do in a repo since it will have to
+// fetch the dependency from the internet.
+func FetchCommonDeps() {
+	// github.com/golang/go
+	d, _ := gosrc.ResolveImportPath(http.DefaultClient, "time")
+	u, _ := url.Parse(d.CloneURL)
+	_, _ = NewDepRepoVFS(context.Background(), u, d.Rev)
+}
+
+// NewDepRepoVFS returns a virtual file system interface for accessing
+// the files in the specified repo at the given commit. The returned VFS
+// is always backed by a zip archive in memory. The following sources are
+// tried in sequence, and the first one that has the repo is used:
+//
+// 1. GitHub's codeload endpoint
+// 2. A full `git clone` followed by `git archive --format=zip <rev>`
+//
+// Source 1 is a performance optimization over cloning the whole repo.
+var NewDepRepoVFS = func(ctx context.Context, cloneURL *url.URL, rev string) (ctxvfs.FileSystem, error) {
```

Comment:
what happened to using Sourcegraph's raw API?

> // 1. Directly from gitserver (will be removed in favor of the raw API soon)

Reply:
The corresponding lang-go Sourcegraph extension constructs the zip URL to the raw API (looks like https://sourcegraph.com/some/repo/name@somerevision/-/raw/) and passes it to go-langserver. go-langserver doesn't really need to know or care that it points to Sourcegraph's raw API - all that matters from its perspective is that it can HTTP GET a zip archive from the URL.

Reply:
Oh right! I see a couple of ways of solving this:

1. Provide go-langserver with a zip URL template (e.g. `https://sourcegraph.com/%s@%s/-/raw`) somehow (command-line args, env vars, LSP `initializationOptions`, etc.). The raw API would need to support clone URLs in order for this to work, otherwise go-langserver will need to convert clone URLs to repo names via Sourcegraph's GraphQL API, which would go against the goal of removing Sourcegraph-specific code from language servers. @slimsag Any thoughts on this?
2. Add a new LSP request method `resolveZipURL` that the go-langserver can send to the Sourcegraph extension asking for a zip URL provided a clone URL.

Any other ideas?

Reply:
I believe a `.netrc` will work for private deps.

I plan to implement option 1 above with the assumption that the repo name is the same as the clone URL, which is usually true. When it's not, the raw API will likely 404 and go-langserver will fall back to a git clone.

Reply:
I implemented option 1 (zip template + HEAD request to the raw API, depends on https://github.com/sourcegraph/sourcegraph/pull/1347).

Reply:
> I plan to implement option 1 above with the assumption that the repo name is the same as the clone URL, which is usually true. When it's not, the raw API will likely 404 and go-langserver will fall back to a git clone.

SGTM

Reply:
I'll document `.netrc` and SSH keys in https://github.com/sourcegraph/lang-go